MobiusCore/Source/EffectHandlers/Response.swift

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2019 Spotify AB.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+
+/// A `Response` can send output and signal completion.
+/// Sending output is done with `.send` and signaling completion is done with `.end`.
+///
+/// Note: Once `.end` has been called (from any thread), the closure you provide in `onSend` will no longer be called.
+/// Note: The closure you provide in `onEnd` will only be called once when `.end` is called on this object.
+public final class Response<T> {
+    private let onSend: (T) -> Void
+    private let onEnd: () -> Void
+
+    public var ended: Bool {
+        return _ended.value
+    }
+
+    private let lock = Lock()
+    private let _ended: Synchronized<Bool> = .init(value: false)
+
+    public init(
+        onSend: @escaping (T) -> Void,
+        onEnd: @escaping () -> Void
+    ) {
+        self.onSend = onSend
+        self.onEnd = onEnd
+    }
+
+    public func end() {
```

Comment:
Would it be problematic for this function to take an optional output?
```swift
public func end(_ finalOutput: T? = nil) {
    …
    if runOnEnd {
        if let output = finalOutput {
            onSend(output)
        }
        onEnd()
    }
}
```
Perhaps the idea was already discussed, but it would make the API usage a little cleaner.

Reply:
Cleaner for the specific case when you want to send a single event, I guess you mean? The drawback, ofc, is that it gives the `end` method two responsibilities as opposed to just one. Also, it creates a difference between this API and the (current) Kotlin proposal, and prior art in the form of the design from reactive-streams.org. The advantage is that it makes one use case a one-liner instead of a two-liner. I don't think that's a huge advantage, but it's definitely an advantage.

Reply:
Correct; "cleaner" in the sense that the response could always be ended in one line, regardless of how many events it sent overall.

In any case, it was based on an initial reaction to the usage examples, so if it breaks API parity and/or clashes with the design principles, those bothered enough by making two function calls could just as easily extend the Response class to wrap them in a single call.